File: backend/api/app/models/proof_of_delivery.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, String, Text, func
from app.database import GUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class ProofOfDelivery(Base):
    __tablename__ = "proofs_of_delivery"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    shipment_id: Mapped[uuid.UUID] = mapped_column(GUID(), ForeignKey("shipments.id"), nullable=False, unique=True)
    signature_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    photo_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    latitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    longitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    recipient_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    driver_id: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("drivers.id"), nullable=True)
    delivered_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

File: packages/shared/lib/src/models/proof_of_delivery.dart
class ProofOfDelivery {
  final String id;
  final String shipmentId;
  final String? signatureUrl;
  final String? photoUrl;
  final double? latitude;
  final double? longitude;
  final String? recipientName;
  final String? notes;
  final String? driverId;
  final DateTime deliveredAt;

  ProofOfDelivery({
    required this.id,
    required this.shipmentId,
    this.signatureUrl,
    this.photoUrl,
    this.latitude,
    this.longitude,
    this.recipientName,
    this.notes,
    this.driverId,
    required this.deliveredAt,
  });

  factory ProofOfDelivery.fromJson(Map<String, dynamic> json) {
    return ProofOfDelivery(
      id: json['id'] as String,
      shipmentId: json['shipment_id'] as String,
      signatureUrl: json['signature_url'] as String?,
      photoUrl: json['photo_url'] as String?,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      recipientName: json['recipient_name'] as String?,
      notes: json['notes'] as String?,
      driverId: json['driver_id'] as String?,
      deliveredAt: DateTime.parse(json['delivered_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'shipment_id': shipmentId,
      'signature_url': signatureUrl,
      'photo_url': photoUrl,
      'latitude': latitude,
      'longitude': longitude,
      'recipient_name': recipientName,
      'notes': notes,
      'driver_id': driverId,
      'delivered_at': deliveredAt.toIso8601String(),
    };
  }
}
